README.md
# Cyberrunner

## SHORT OVERVIEW
Cyberrunner is a mod for the ZDoom/Zandronum engine that essentially acts as a combination of ijon tichy's Parkmore (which you probably have heard of) and TerminusEst13's Superfast (which you probably haven't), taking the best of both and turning it into a full-fledged mod. In Cyberrunner, players take control of powerful and fast cyborgs/robots with large laser weapons, dodging and weaving each other's attacks, the environment, and the enemies within in order to get to the finish as quickly as possible.

## PACKAGING AND PLAYING
If you aren't a Cyberrunner developer and want to see what new unreleased shit there is, this is all you need to do:

 - Click the "ZIP" button below the Cyberrunner summary.
 - Unzip everything in the resulting ZIP file to an empty folder.
 - Navigate into the pk3/ folder.
 - Select everything (ctrl-A).
 - Put it all in a ZIP or 7-Zip (7z) file. Do *not* use RAR, since Zandronum does not support most versions of RAR.
 - Change the name of the ZIP file to "cyberrunner-git.pk3".
 - Play Cyberrunner.

If you're a developer, the process is the same, but remember to recompile ACS when you change it. ACC must be in the pk3/acs/ directory when you recompile, or else it'll blow up on the #includes (if this is wrong, tell ijon on irc.zandronum.com/#cyberrunner).

## CONTACT
TerminusEst13 and various other Cyberrunner players/mappers are on #cyberrunner on irc.zandronum.com. You'll need an IRC client to get there; look at HexChat ( http://hexchat.org/ ) if you're on Windows, or XChat if you're on Linux.


## MAPPER INFO
If you want to map, there's a few simple things to keep in mind.

- This is CYBERrunner, so it should be at least vaguely sci-fi in theme. Set in our solar system; outer space, the asteroid field, in a spatial anomaly, on Neptune, an abandoned space station, haunted ghost ship, whatever. It's generally pretty flexible, as long as it's sci-fi.
- Use Doom in Hexen or UDMF format for your maps. Cyberrunner makes use of various advanced features for mappan.
- Maps should be very wide open and spacious. This doesn't mean make a 10000x10000 field, but it should be very possible to move through at high speeds.
- Cyberrunner uses tracker music. That means .s3m, .it, .xm, .mod, .spc, and .vgz music. This provides the wonderful synthesized instruments so reminiscent of retro sci-fi, while being VERY easy on filesize, AND gives us the wonderful ability to edit it as we please. Sorry, but we're not going to use MP3s. If you don't know of any good tracker music, let me know and I can help.
- Custom textures and flats are allowed. If you see a set that just looks absolutely awesome, go for it! We don't really have any limit for how many you can use, but please don't suddenly implement 80 megs of textures.
- Criticism is inevitable. Nobody gets everything perfect on the first go! Please be able to take criticism--or failing that, please understand that your map may be edited and tweaked.

If you want to have checkpoints in your map, use script 105. It is defined like this:

  script 105 (int mode, int index, int nextToBe)

The mode argument works like this:
 - Mode 0 teleports the player to the coordinates of their checkpoint.
 - Mode 1 sets their checkpoint. This will not happen instantly, but rather, it'll happen when they hit the ground. When they hit the ground, their position, angle, and pitch will be saved.
 - Mode 2 is the same as mode 1, but silent.
 - Modes 3 and 4 are the same as 1 and 2, respectively, but will always set a checkpoint.
 - Mode 5 is the finishline! It's the big finisher that activates sudden death.

The index argument is the number of the checkpoint. 105,1,1 is the first checkpoint, 105,1,2 is the second checkpoint, etc. Simple enough.

For most intents and purposes, you can keep nextToBe set to 0. Starting your checkpoints at index 1 is suggested, and if you have a track that loops, set the "lap line" to checkpoint 0. If you do a lapping map, you must use checkpoints, even if you don't then teleport the player to those checkpoints.


## JUST HOW FAST CAN THE RUNNER GO, ANYWAY?

We've calculated it to be 135494.272 mph. 0.1% away from Mach 178.

So pretty goddamn fast.


changelog.txt
FROM 0.11 TO 0.12:
===================
- The Cyber Claw now makes its appearance! A laser grapple hook with a range of 4072 units, it latches onto anything it hits and starts pulling the Runner in, allowing them to yank themselves to new areas or simply swing along. It is an optional weapon, activated via cvar of cyber_claw



FROM 0.1 TO 0.11:
===================
- Added a new Deathmatch map, CYDM03: Vertical Datapath, by Revae!
- Trimmed a few unused textures.
- Rephrased some obituaries due to incorrect/lame terminology.
- The Shotgun's tracers now show up in non-Doom games.
- Weapon balance has been completely redone using the power of MATH, rather than "what feels right".
 ^-: Carbine now does 180 damage every 50 tics. 180/50 = 3.6, 3.6*35 = average 126 damage a second. 21 damage buff.
 ^-: Shotgun does 10*4(1d3) and 15*3(1d3), 80+90 = 170, 170/40 = 4.25, 4.25*35 = average 148.75 damage a second. 8.75 damage buff.
 ^-: Vulcan does two projectiles with 8 damage every 3 tics. [(2x8)/2]+(8/2) = 12 damage every 3 tics = 4, 4*35 = average 140 damage a second. 35 damage nerf.
 ^-: Vulcan has had its ammo halved in LMS, with 3 tics at 500 ammo being 1500/35 = 42.85[...] seconds of firing. Its radius and height has also been reduced from 8 to 6.
 ^-: Running people over has had its damage boosted from 400 to 500, to ensure sure-kills.
- Running people over no longer can be reflected back and kill the Runner. That was weird.
- Ammo numbers for the Shotgun, Carbine, Purge Rifle, and Hammer have been centered in the bar HUD.
- The Cyberrunner tutorial, racing maps, and deathmatch maps are now accessible from the main menu.
- There is now a Quake-style when-shot-connects sound indicator, toggled via cyber_cl_hitindicator.
- Various tweaks across both CYRC01 and CYRC02, adjusting textures and tweaking the ACS.
- The race maps now refill health and boosts upon enter.
- cyber_cl_particles has been removed, in an attempt to combat the crashing that some people are having.
- For vanilla jumping maps with combat, ammo drops have been added for the weapons. They replace the big ammo packs and have about a quarter of the weapon's total ammo capacity.
- Nothing has changed in CYRTUT.
- With Zandronum's update to 1.1, the Shotgun now has FBF_NORANDOM for normalized damage.
- CYDM02 has had the side nodes removed.
- CYRC02 has been rendered more open for easier navigation, with temporary additional structures for easier traversing until it can be put later in the map rotation.
- cyber_changelog has been implemented.

pk3/acs/cyberdefs.h
#define MODES_OPEN          511
#define MODES_ENTER         512
#define MODES_DISCONNECT    513
#define MODES_SWITCH        514

#define MODES_OPEN_CLIENT   531
#define MODES_ENTER_CLIENT  522
#define MODES_INFORM        523

#define ENDGAME_SCOREBOARD  515

#define TID_PLAY 600

#define DASH_VEL 45

#define TIMECOUNT   3
#define TIME_START      0
#define TIME_CHECKPOINT 1
#define TIME_END        2
#define TIME_FINISH     2

#define PLACEMAX 64

int PlayerTIDs[PLAYERMAX];
int PlayerTimes[PLAYERMAX][TIMECOUNT];
int InTerminal[PLAYERMAX];
int UnfreezeDelay[PLAYERMAX];
int IsServer = 0;
int TimeDisplays[PLACEMAX][3] = {{-1, -1, -1}};  // name pointer, time, player number
int PlayerPlace[PLAYERMAX] = {-1}; 
int ReadyToExit[PLAYERMAX] = {0};
int Purged[PLAYERMAX] = {0};

#define BARFONTCOUNT 12

int BarGraphics[BARFONTCOUNT] = // These are the different graphics that make up the MPH bar.
{
    "MPHBAR1", "MPHBAR2", "MPHBAR3", "MPHBAR4", "MPHBAR5", "MPHBAR6",
    "MPHBAR7", "MPHBAR8", "MPHBAR9", "MPHBAR10", "MPHBAR11", "MPHBAR12",
};

#define RECHARGECOUNT 11

int RechargingItems[RECHARGECOUNT] = // These are the items that "recharge" over time, that are slowly removed or added to the player's inventory.
{
    "DashCooldown", "JumpCooldown", "BoostCooldown", "PlasmaGunAmmo",
    "ForceVentAmmo", "CannotIntoShotgun", "CannotIntoCarbine",
    "CannotIntoVulcan", "RedPlasmaGunAmmo", "CannotIntoSuper", "CyberClawAmmo"
};

int RechargingTimes[RECHARGECOUNT][2] = // The max amount of an item, and whether 1 is given or -1 is taken.
{
    {5, -1}, {5, -1}, {5, -1}, {15, 1}, {210, 1}, {1, -1}, {1, -1}, {1, -1}, {15, 1}, {1, -1}, {140, 1}
};

#define RAINBOWCOLORS 12
#define RAINBOWINTERVAL 10

int RainbowBeamActors[RAINBOWCOLORS] = // The Purge Rifle's pretty colors.
{
    "RainbowBeam1",
    "RainbowBeam2",
    "RainbowBeam3",
    "RainbowBeam4",
    "RainbowBeam5",
    "RainbowBeam6",
    "RainbowBeam7",
    "RainbowBeam8",
    "RainbowBeam9",
    "RainbowBeam10",
    "RainbowBeam11",
    "RainbowBeam12",
};

#define PARTICLECOUNT 12
#define PARTDIST_MIN  8.0
#define PARTDIST_MAX  32.0
 
int ParticleTypes[PARTICLECOUNT] = // The particles that are summoned from an item
{
    "CyberShotgunSparkle", "CyberCarbineSparkle", "CyberVulcanSparkle", "ForceVentSparkle",
    "TurboSparkle", "AllAmmoSparkle", "50HPSparkle", "InstagibSparkle", "25HPSparkle",
	"RedSparkle", "WhiteSparkle", "BlueSparkle",
};

#define PLACENAMECOUNT 10

int PlaceNames[PLACENAMECOUNT] = // The ranks on a scoreboard
{
    "\cfFirst",
    "\cjSecond",
    "\ceThird",
    "\cdFourth",
    "\caFifth",
    "\cnSixth",
    "\cpSeventh",
    "\ctEighth",
    "\cuNinth",
    "\csTenth",
};

// 0, 1, 2  -> x, y, z
// 3, 4     -> angle, pitch
// 5        -> last checkpoint
// 6        -> next checkpoint
int CheckpointCoords[PLAYERMAX][8];
int OldCheckpointCoords[PLAYERMAX][8];
int HasTeleported[PLAYERMAX];

#define MODECOUNT   6

#define MODE_NORMAL         1
#define MODE_COUNTDOWN      2
#define MODE_SUDDENDEATH    3
#define MODE_PURGE          4
#define MODE_SCOREBOARD     5
#define MODE_ABORT          6

int CRGameMode = MODE_NORMAL;
int CRSwitchTime = 0;
int CRSwitchTo = -1;
int CRSwitchLock;
int CRSwitched = 0;

int SuddenDeathEnd = 0;

int ClientPlace;

int CRModeNames[MODECOUNT+1] = // These are the modes activated one after each other in the finish sequence
{
    "GrossHack", // <ijon> the "GrossHack" is there so I don't have to subtract 1 every time I want to get the name of a CR mode
    "Normal", // Playing normally
    "Countdown", // Hit the finish line, time to countdown
    "Sudden Death", // Mid-Sudden Death
    "Failure Purge", // Didn't make it to the finish line, everyone gets kabloom'd
    "Scoreboard", // Display the scoreboard
    "Sudden Death Abort", // Player who hit the switch left the game
};

int ChangelogCyber =
"FROM V0.1 TO V0.11:\n\
========================\n\
- Added a new Deathmatch map, CYDM03: Vertical Datapath, by Revae!\n\
- Trimmed a few unused textures.\n\
- Rephrased some obituaries due to incorrect/lame terminology.\n\
- The Shotgun's tracers now show up in non-Doom games.\n\
- Weapon balance has been completely redone using the power of MATH, rather than 'what feels right'.\n\
 ^-: Carbine now does 180 damage every 50 tics. 180/50 = 3.6, 3.6*35 = average 126 damage a second. 21 damage buff.\n\
 ^-: Shotgun does 10*4(1d3) and 15*3(1d3), 80+90 = 170, 170/40 = 4.25, 4.25*35 = average 148.75 damage a second. 8.75 damage buff.\n\
 ^-: Vulcan does two projectiles with 8 damage every 3 tics. [(2x8)/2]+(8/2) = 12 damage every 3 tics = 4, 4*35 = average 140 damage a second. 35 damage nerf.\n\
 ^-: Vulcan has had its ammo halved in LMS, with 3 tics at 500 ammo being 1500/35 = 42.85[...] seconds of firing. Its radius and height has also been reduced from 8 to 6.\n\
 ^-: Running people over has had its damage boosted from 400 to 500, to ensure sure-kills.\n\
- Running people over no longer can be reflected back and kill the Runner. That was weird.\n\
- Ammo numbers for the Shotgun, Carbine, Purge Rifle, and Hammer have been centered in the bar HUD.\n\
- The Cyberrunner tutorial, racing maps, and deathmatch maps are now accessible from the main menu.\n\
- There is now a Quake-style when-shot-connects sound indicator, toggled via cyber_cl_hitindicator.\n\
- Various tweaks across both CYRC01 and CYRC02, adjusting textures and tweaking the ACS.\n\
- The race maps now refill health and boosts upon enter.\n\
- cyber_cl_particles has been removed, in an attempt to combat the crashing that some people are having.\n\
- For vanilla jumping maps with combat, ammo drops have been added for the weapons. They replace the big ammo packs and have about a quarter of the weapon's total ammo capacity.\n\
- Nothing has changed in CYRTUT.\n\
- With Zandronum's update to 1.1, the Shotgun now has FBF_NORANDOM for normalized damage.\n\
- CYDM02 has had the side nodes removed.\n\
- CYRC02 has been rendered more open for easier navigation, with temporary additional structures for easier traversing until it can be put later in the map rotation.";